---
layout: splash
title: About
permalink: /about/
---

<img align="right" src="https://avatars1.githubusercontent.com/u/862013" width=200px height=200px>

Hi! I am a software engineer and a bioinformatician. I work at [OpenAI](https://openai.com). Previously I worked at
[Color](https://color.com), where I was lucky to help build the future of genetic disease risk modeling, population
genomics, and healthcare delivery, as well as provide emergency diagnostic services during the COVID-19 pandemic.

Prior to Color, I worked at the
[Chan Zuckerberg Initiative](https://chanzuckerberg.com/science/), where I helped build
infrastructure for the [CZID](https://czid.org/) project (also known as IDseq); a lot of this work is available
as open source in the [SWIPE](https://github.com/chanzuckerberg/swipe) and 
[czid-workflows](https://github.com/chanzuckerberg/czid-workflows) repositories.

Prior to CZI, I was involved with several biotechnology startups which gave me the opportunity to become an expert in
the design and implementation of genome analysis systems and cloud platforms. I am particularly known for building
state of the art systems for DNA sequence analysis, machine learning applications in genomics, scalable distributed
systems for genomics, API design, Linux systems, information security, and development operations. Specific applications
that I was involved with include single-molecule DNA sequencing R&D, microbial genome and metagenome annotation and
interpretation, single-cell sequencing, variant calling/interpretation, and phylogenomics. Since 2011, I have
been using Amazon Web Services to build these solutions, and I am now an expert in many AWS technologies as well as
Terraform, an essential [IaC](https://en.wikipedia.org/wiki/Infrastructure_as_code) tool used to configure and manage
assets on AWS. Some of the tools my teams use every day to manage AWS assets are available as open source in the
[Aegea](https://github.com/kislyuk/aegea) project, as well as other projects listed on my
[GitHub profile](https://github.com/kislyuk).

Before my industry career, I graduated from [UC Berkeley](https://www.berkeley.edu/) with a triple major in computer
science, mathematics, and statistics. I then went to grad school at [Georgia Tech](https://www.gatech.edu/) and
graduated with a PhD in bioinformatics. My undergraduate research work was at Lawrence Berkeley Lab with
[Inna Dubchak](https://biosciences.lbl.gov/profiles/inna-dubchak-2/) and
[Mike Brudno](http://www.cs.toronto.edu/~brudno/public/). Together with Mike's advisor
[Serafim Batzoglou](https://en.wikipedia.org/wiki/Serafim_Batzoglou), we created the
[first global multiple genome alignment of multiple large eukaryotic genomes](https://genome.cshlp.org/content/19/4/682.long).
My graduate advisor was [Joshua Weitz](https://en.wikipedia.org/wiki/Joshua_Weitz), and my thesis title was
[Algorithm development for next generation sequencing-based metagenome analysis](https://smartech.gatech.edu/bitstream/handle/1853/42779/kislyuk_andrey_o_201012_phd.pdf).
You can read more exciting details in my [CV](akislyuk.pdf). But a lot of what I learned during my time in grad school,
I actually learned at [Black Knight Martial Arts](http://www.blackknight.com/).

A few notable state of the art infrastructure projects that I implemented include an
[LXC](https://linuxcontainers.org/)-based cloud PaaS virtualization service which continues to power millions of jobs
and Docker apps on the [DNAnexus Titan](https://www.dnanexus.com/product-overview/titan) platform; an independent
implementation of [SAML](https://en.wikipedia.org/wiki/Security_Assertion_Markup_Language), OAuth2/OIDC, and their
underlying technologies for single sign-on applications; an AWS IAM-based symbolic RBAC PDP service; and multiple API
designs, CLI tools, and developer productivity tools for a variety of products. On the science side, I have mostly
been involved in integration and tuning of existing genomics and machine learning software written by people smarter
than me, but you can see some of my concrete contributions on my [GitHub](https://github.com/kislyuk) and
[Google Scholar](https://scholar.google.com/citations?user=7deTos4AAAAJ&sortby=pubdate) profiles.

The technologies I use in my daily work include Terraform, AWS Fargate, S3, EFS, Lambda, Step Functions, IAM,
RDS/Aurora, Python, Rust, OIDC, LXC, Docker, [cloud-init](https://cloud-init.io/), as well as many genomics tools
including [Kraken2](https://ccb.jhu.edu/software/kraken2/), [miniwdl](https://github.com/chanzuckerberg/miniwdl),
[bedtools2](https://github.com/arq5x/bedtools2), and [minimap2](https://github.com/lh3/minimap2).

I live in San Francisco. In my spare time I do a lot of the usual suspect activities such as biking, running,
photography, and selecting the proper power tools for home improvement.
